File: node/src/runtime/archive.ts
import { mkdir, writeFile, readFile } from "node:fs/promises"
import { join } from "node:path"
import type { ModelMessage } from "../types.js"

export interface ArchiveStore {
  write(sessionId: string, seq: number, messages: ModelMessage[]): Promise<string>
  read(archiveRef: string): Promise<ModelMessage[]>
}

export class NullArchiveStore implements ArchiveStore {
  async write(_sessionId: string, _seq: number, _messages: ModelMessage[]): Promise<string> {
    return ""
  }

  async read(_archiveRef: string): Promise<ModelMessage[]> {
    throw new Error("NullArchiveStore does not store archives")
  }
}

export class FileArchiveStore implements ArchiveStore {
  constructor(private readonly root: string) {}

  async write(sessionId: string, seq: number, messages: ModelMessage[]): Promise<string> {
    const dir = join(this.root, sessionId)
    await mkdir(dir, { recursive: true })
    const filePath = join(dir, `${seq}.jsonl`)
    const lines = messages.map(msg => JSON.stringify(msg)).join("\n") + "\n"
    await writeFile(filePath, lines, "utf8")
    return filePath
  }

  async read(archiveRef: string): Promise<ModelMessage[]> {
    const content = await readFile(archiveRef, "utf8")
    const lines = content.split("\n")
    const messages: ModelMessage[] = []
    for (const line of lines) {
      if (!line.trim()) continue
      messages.push(JSON.parse(line) as ModelMessage)
    }
    return messages
  }
}

File: rust/src/runtime/archive.rs
use deepstrike_core::types::message::Message;
use std::error::Error;
use std::fs::{self, File};
use std::io::Write;
use std::path::PathBuf;

pub trait ArchiveStore: Send + Sync {
    fn write(
        &self,
        session_id: &str,
        seq: u64,
        messages: &[Message],
    ) -> Result<String, Box<dyn Error + Send + Sync>>;
}

pub struct NullArchiveStore;

impl ArchiveStore for NullArchiveStore {
    fn write(
        &self,
        _session_id: &str,
        _seq: u64,
        _messages: &[Message],
    ) -> Result<String, Box<dyn Error + Send + Sync>> {
        Ok(String::new())
    }
}

pub struct FileArchiveStore {
    pub root: PathBuf,
}

impl FileArchiveStore {
    pub fn new(root: PathBuf) -> Self {
        Self { root }
    }
}

impl ArchiveStore for FileArchiveStore {
    fn write(
        &self,
        session_id: &str,
        seq: u64,
        messages: &[Message],
    ) -> Result<String, Box<dyn Error + Send + Sync>> {
        let dir = self.root.join(session_id);
        fs::create_dir_all(&dir)?;
        let file_path = dir.join(format!("{}.jsonl", seq));
        let mut file = File::create(&file_path)?;

        for msg in messages {
            let line = serde_json::to_string(msg)?;
            writeln!(file, "{}", line)?;
        }

        Ok(file_path.to_string_lossy().to_string())
    }
}
